File: netlify/functions/recipe-detail.py
"""
Netlify Function: Recipe Detail API
Handles GET (single recipe), PUT (update), DELETE (delete)
"""
import json
from utils.db import execute_query, execute_update, execute_delete, format_response, handle_error

def handler(event, context):
    """Main handler for recipe detail endpoint"""

    # Handle CORS preflight
    if event['httpMethod'] == 'OPTIONS':
        return format_response(200, {})

    try:
        # Extract recipe ID from path
        path = event.get('path', '')
        recipe_id = path.split('/')[-1]

        if not recipe_id or not recipe_id.isdigit():
            return format_response(400, {
                "error": "Invalid recipe ID",
                "success": False
            })

        recipe_id = int(recipe_id)

        # GET: Get single recipe
        if event['httpMethod'] == 'GET':
            return get_recipe(recipe_id)

        # PUT: Update recipe
        elif event['httpMethod'] == 'PUT':
            return update_recipe(recipe_id, event)

        # DELETE: Delete recipe
        elif event['httpMethod'] == 'DELETE':
            return delete_recipe(recipe_id)

        else:
            return format_response(405, {
                "error": "Method not allowed",
                "success": False
            })

    except Exception as e:
        return handle_error(e)

def get_recipe(recipe_id):
    """Get a single recipe by ID"""
    try:
        query = "SELECT * FROM recipes WHERE id = %s"
        recipe = execute_query(query, (recipe_id,), fetch='one')

        if not recipe:
            return format_response(404, {
                "error": "Recette non trouvée",
                "success": False
            })

        return format_response(200, {
            "success": True,
            "recipe": recipe
        })

    except Exception as e:
        return handle_error(e)

def update_recipe(recipe_id, event):
    """Update an existing recipe"""
    try:
        # Parse request body
        body = json.loads(event['body'])

        # Check if recipe exists
        check_query = "SELECT id FROM recipes WHERE id = %s"
        existing = execute_query(check_query, (recipe_id,), fetch='one')
        if not existing:
            return format_response(404, {
                "error": "Recette non trouvée",
                "success": False
            })

        # Check for duplicate title (if title is being changed)
        if 'titre' in body:
            dup_query = "SELECT id FROM recipes WHERE titre = %s AND id != %s"
            duplicate = execute_query(dup_query, (body['titre'], recipe_id), fetch='one')
            if duplicate:
                return format_response(409, {
                    "error": "Une autre recette avec ce titre existe déjà",
                    "success": False
                })

        # Prepare update data (only include fields that are present)
        allowed_fields = ['titre', 'temps_preparation', 'temps_cuisson', 'rendement',
                         'ingredients', 'instructions', 'image_url', 'category', 'tags']

        update_fields = []
        update_values = []

        for field in allowed_fields:
            if field in body:
                update_fields.append(f"{field} = %s")
                # Convert lists/dicts to JSON strings for JSONB fields
                if field in ['ingredients', 'instructions']:
                    update_values.append(json.dumps(body[field]))
                else:
                    update_values.append(body[field])

        if not update_fields:
            return format_response(400, {
                "error": "No fields to update",
                "success": False
            })

        # Build and execute update query
        update_query = f"""
            UPDATE recipes
            SET {', '.join(update_fields)}
            WHERE id = %s
            RETURNING *
        """
        update_values.append(recipe_id)

        recipe = execute_update(update_query, tuple(update_values))

        return format_response(200, {
            "success": True,
            "recipe": recipe,
            "message": "Recette mise à jour avec succès"
        })

    except json.JSONDecodeError:
        return format_response(400, {
            "error": "Invalid JSON in request body",
            "success": False
        })
    except Exception as e:
        return handle_error(e)

def delete_recipe_handler(recipe_id):
    """Delete a recipe"""
    try:
        # Check if recipe exists
        check_query = "SELECT id FROM recipes WHERE id = %s"
        existing = execute_query(check_query, (recipe_id,), fetch='one')
        if not existing:
            return format_response(404, {
                "error": "Recette non trouvée",
                "success": False
            })

        # Delete recipe
        delete_query = "DELETE FROM recipes WHERE id = %s"
        execute_delete(delete_query, (recipe_id,))

        return format_response(200, {
            "success": True,
            "message": "Recette supprimée avec succès",
            "rows_deleted": rows_deleted
        })

    except Exception as e:
        return handle_error(e)

File: netlify/functions/recipe-detail.mjs
import { json, err } from "./lib/http.mjs";
import { queryOne, execReturning, execCommand } from "./lib/db.mjs";

function recipeIdFromPath(event) {
  const path = event.path || event.rawPath || "";
  const last = path.split("/").filter(Boolean).pop() || "";
  return last;
}

export const handler = async (event) => {
  if (event.httpMethod === "OPTIONS") return json(200, {});

  const rawId = recipeIdFromPath(event);
  if (!rawId || !/^\d+$/.test(rawId)) {
    return err(400, "Invalid recipe ID");
  }
  const recipeId = parseInt(rawId, 10);

  try {
    if (event.httpMethod === "GET") {
      const recipe = await queryOne("SELECT * FROM recipes WHERE id = $1", [recipeId]);
      if (!recipe) return err(404, "Recette non trouvée");
      return json(200, { success: true, recipe });
    }

    if (event.httpMethod === "PUT") {
      let body;
      try {
        body = JSON.parse(event.body || "{}");
      } catch {
        return err(400, "Invalid JSON in request body");
      }

      const existing = await queryOne("SELECT id FROM recipes WHERE id = $1", [recipeId]);
      if (!existing) return err(404, "Recette non trouvée");

      if (body.titre != null) {
        const duplicate = await queryOne(
          "SELECT id FROM recipes WHERE titre = $1 AND id != $2",
          [body.titre, recipeId]
        );
        if (duplicate) {
          return err(409, "Une autre recette avec ce titre existe déjà");
        }
      }

      const allowed = [
        "titre",
        "temps_preparation",
        "temps_cuisson",
        "rendement",
        "ingredients",
        "instructions",
        "image_url",
        "category",
        "tags",
        "source_url",
      ];

      const sets = [];
      const args = [];
      for (const field of allowed) {
        if (!(field in body)) continue;
        args.push(
          field === "ingredients" || field === "instructions"
            ? JSON.stringify(body[field])
            : body[field]
        );
        const cast =
          field === "ingredients" || field === "instructions" ? "::jsonb" : "";
        sets.push(`${field} = $${args.length}${cast}`);
      }

      if (sets.length === 0) return err(400, "No fields to update");

      args.push(recipeId);
      const sql = `UPDATE recipes SET ${sets.join(", ")} WHERE id = $${args.length} RETURNING *`;
      const recipe = await execReturning(sql, args);
      return json(200, {
        success: true,
        recipe,
        message: "Recette mise à jour avec succès",
      });
    }

    if (event.httpMethod === "DELETE") {
      const existing = await queryOne("SELECT id FROM recipes WHERE id = $1", [recipeId]);
      if (!existing) return err(404, "Recette non trouvée");

      const rows = await execCommand("DELETE FROM recipes WHERE id = $1", [recipeId]);
      return json(200, {
        success: true,
        message: "Recette supprimée avec succès",
        rows_deleted: rows,
      });
    }

    return err(405, "Method not allowed");
  } catch (e) {
    console.error(e);
    return json(500, { success: false, error: "An internal error occurred" });
  }
};
